C/common/management_client.cpp

```diff
@@ -864,3 +867,308 @@ bool ManagementClient::addAuditEntry(const std::string& code,
+			// Token verified, store the token
+			m_received_tokens[bearer_token] = "added";
+		}
+		else
+		{
+			claims.clear();
+			ret = false;
+			m_logger->error("Micro service bearer token '%s' not verified.",
+					bearer_token.c_str());
+		}
+	}
+	else
+	{
+		// A validated token exists: check expiration time
+		vector<string> elems = JWTTokenSplit(bearer_token, '.');
+		// Add missing base64 padding
+		if (elems[1].length() % 4 != 0)
+		{
+			elems[1] += "=";
+		}
+		if (elems[1].length() % 4 != 0)
+		{
+			elems[1] += "=";
+		}
+		// Base64 decode of second part of bearer token (with public claims)
+		string plainData = Crypto::Base64::decode(elems[1]);
+		Document doc;
+		doc.Parse(plainData.c_str());
+
+		// JSON parsing check of token claims
+		if (doc.HasParseError())
+		{
+			m_logger->error("Failure while parsing JSON claims of bearer token, "
+					"error %s, input data %s",
+					GetParseError_En(doc.GetParseError()),
+					plainData.c_str());
+			claims.clear();
+			return false;
+		}
+
+		unsigned long expiration;
+		if (doc["exp"].IsUint())
+		{
+			expiration = doc["exp"].GetUint();
+		}
+		else
+		{
+			m_logger->error("Failure while parsing 'exp' token claim of bearer token, "
+					"error %s, input data %s",
+					"not a number",
+					plainData.c_str());
+			claims.clear();
+			return false;
+		}
+		unsigned long now = time(NULL);
+
+		// Check expiration
+		if (now >= expiration)
+		{
+			ret = false;
+			// Remove token from received ones
+			m_received_tokens.erase(bearer_token);
+
+			m_logger->error("Micro service bearer token '%s' has expired.",
+					bearer_token.c_str());
+			claims.clear();
+		}
+		else
+		{
+			// Token still valid, set token claims in the input map
+			claims["aud"] = doc["aud"].GetString();
+			claims["sub"] = doc["sub"].GetString();
+			claims["iss"] = doc["iss"].GetString();
+			claims["exp"] = std::to_string(doc["exp"].GetUint());
+		}
+	}
+
+	// Release lock
+	m_mtx_rTokens.unlock();
+
+	m_logger->error("Token verified %d, claims %s:%s:%s:%s",
+			ret,
+			claims["aud"].c_str(),
+			claims["sub"].c_str(),
+			claims["iss"].c_str(),
+			claims["exp"].c_str());
+
+	return ret;
+}
+
+/**
+ * Refresh the JWT bearer token coming from HTTP request
+ *
+ * @param request       HTTP request object
+ * @param claims        Map to fill with JWT public token claims
+ * @return              New bearer token on success, empty string otherwise
+ */
+string ManagementClient::refreshAccessBearerToken(shared_ptr<HttpServer::Request> request)
+{
+	string bearer_token = getAccessBearerToken(request);
+	if (bearer_token.length() == 0)
+	{
+		return "";
+	}
+
+	return refreshBearerToken(bearer_token);
+}
+
+/**
+ * Refresh the JWT bearer token string
+ *
+ * @param request       HTTP request object
+ * @param claims        Map to fill with JWT public token claims
+ * @return              New bearer token on success, empty string otherwise
+ */
+string ManagementClient::refreshBearerToken(const string& bearer_token)
```

Comment:
There are lots of copies going on here, because of object returns. Would it be better to pass std::string&  and avoid the copies. You could then return a boolean status and not have the look for empty string to detect failure.